Function to encode strings

miguelleitao opened the pull request:
Added a function to encode strings.

miguelleitao committed:
encoding

miguelleitao committed:
encoding

miguelleitao committed:
stdio added

miguelleitao committed:
stdio added

miguelleitao committed:
encode declaration

miguelleitao committed:
encode declaration

ooxi commented:
Thanks for your contribution @miguelleitao! I'm very interested in adding this functionality, since we do not have an encoding functionality yet. Please see my inline comments on your merge requests for details (will follow in a couple of minutes :-)

ooxi commented on entities.c:
```diff
@@ -390,3 +390,32 @@ size_t decode_html_entities_utf8(char *dest, const char *src)
 	return (size_t)(to - dest);
 }
 
+int encode_html_entities(char *dest, const char *src) {
```
Please return `sitze_t` instead of `int`

miguelleitao replied:
Completely agree. I'll commit the correction.

ooxi commented on entities.c:
```diff
@@ -390,3 +390,32 @@ size_t decode_html_entities_utf8(char *dest, const char *src)
+int encode_html_entities(char *dest, const char *src) {
+        char *to = dest;
+        for( const char *from = src ; *from ; from++ ) {
+            int i = 9999;
```
Please declare `i` as late as possible and do not assign a magic unused value to it

miguelleitao replied:
Completely agree. I'll commit the correction.

ooxi commented on entities.c:
```diff
@@ -390,3 +390,32 @@ size_t decode_html_entities_utf8(char *dest, const char *src)
+            }
+            //if ( *from=='\r' || *from=='\n' ) continue;
+            for( i=0 ; i<sizeof NAMED_ENTITIES / sizeof *NAMED_ENTITIES ; i++ )
+                if ( *from == NAMED_ENTITIES[i][1][0] ) break;
```
I don't really get this logic :-) are you comparing just the first character or am I missing something? Or is this sufficient because of the available entities (then please make sure this remains the case by adding an appropriate test case)

miguelleitao replied:
Yes, I'm just comparing the first character. I found this sufficient for the final application I was developing. Maybe this encoder should be classified as an ASCII encoder instead of an UTF-8 encoder.  I'll think about this during next week. Enlarging the length of the test can be done but this solution will still be inefficient since it uses a top-down search on an unsorted table.
If I won't improve the test, I'll try to keep the restriction clear to future users or developers.
Any way, I'll provide a test case.

ooxi commented on entities.c:
```diff
@@ -390,3 +390,32 @@ size_t decode_html_entities_utf8(char *dest, const char *src)
+            }
+        }
+        *to = 0;
+        return strlen(dest);
```
Since you know `dest` and `to`, no call to `strlen` is necessary

miguelleitao replied:
Completely agree. I'll commit the correction.

ooxi reviewed (changes requested):
Needs some work, but looks promising! Please also add a test case for this functionality.

miguelleitao committed:
size_t return

miguelleitao committed:
size_t return

miguelleitao committed:
const cast in cmp() to improve compatibility

miguelleitao committed:
encoding with full test

miguelleitao committed:
encoding with full test

miguelleitao committed:
added commat; entitie

miguelleitao committed:
encoding test

miguelleitao committed:
encoding test

miguelleitao commented:
Hi ooxi. Thank you for your reviews. I just pushed a new version of the encoding function.
I implemented the character matching test using all the UTF bytes and included a simple test in t-entities.c.
I removed the previously included percent coding because I believe that this solution should be closer to the expected from this package. When required, the percent coding can be implemented from another package or a developed function. This way, the encoding function should produce the opposite result to the already provided decoding function.
Remaining issues: 
- the encoding function uses top-down search which is not efficient.
- the result is unspecified when the character is not found in the table.
Hope you find this simple implementation usable.
Let me know of any questions.

ooxi commented:
Thanks for splitting your changes into topic commits, of which I already have merged several into master! The rest can be found in the [feature/encode](https://github.com/ooxi/entities/tree/feature/encode) branch and will be merged as soon as I have had a chance to look at it :-)

ooxi committed:
Misc stylistic unification

miguelleitao committed:
Merge branch 'feature/encode' of https://github.com/ooxi/entities

gjtorikian commented:
> Thanks for splitting your changes into topic commits, of which I already have merged several into master! The rest can be found in the [feature/encode](https://github.com/ooxi/entities/tree/feature/encode) branch and will be merged as soon as I have had a chance to look at it :-)

Was this ever merged? 😅

ooxi commented:
@gjtorikian unfortunately, no. But feel free to use it as is :-)

miguelleitao committed:
Merge branch 'ooxi:master' into master
